src/tools/miri/src/machine.rs

```diff
@@ -967,9 +972,26 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for MiriMachine<'mir, 'tcx> {
             )
         });
         let buffer_alloc = ecx.machine.weak_memory.then(weak_memory::AllocState::new_allocation);
+
+        // If an allocation is leaked, we want to report a backtrace to indicate where it was
+        // allocated. We don't need to record a backtrace for allocations which are allowed to
+        // leak.
+        let backtrace = if kind.may_leak() {
+            None
+        } else {
+            let backtrace = ecx.generate_stacktrace();
+            let (backtrace, _pruned) = prune_stacktrace(backtrace, &ecx.machine);
+            Some(backtrace)
+        };
```

Comment:
We do not have a benchmark. But on this program:
```
fn main() {
    for _ in 0..10_000 {
        let _ = Box::new(0u8);
    }
}
```
with this PR as-written, we spend about 50% of our time in `prune_backtrace`.

The implementation I wrote initially, which did not call `ecx.generate_stacktrace` and `prune_stacktrace`, was designed to avoid this overhead.

I think we should make the implementation efficient instead of offering people a chance to turn off the debugging support, which is a decision they will most regret when they need it the most. I'll try to make this efficient enough that it is tolerable.